page.go

```diff
@@ -40,32 +40,36 @@ type page struct {
 	// pageStatus is set to '4' if the transaction has been committed and
 	// applied, meaning that the transaction can be ignored upon load, and the
 	// associated pages can be reclaimed.
-	pageStatus uint64 // Gets marshalled to disk.
+	pageStatus uint64 // marshalled to disk.
 
 	// nextPage points to the logical next page in the logFile for this
 	// transaction. The page may not appear in the file in-order. This value
 	// is set to math.MaxUint64 if there is no next page (indicating this
 	// page is the last). The first page may be the only page, in which
 	// case it is also the last page.
-	nextPage *page // Gets marshalled as nextPage.offset.
+	nextPage *page // marshalled as nextPage.offset.
 
 	// transactionNumber is only saved for the last page, which can be
 	// determined by nextPage being nil, or when marshalled to disk if
 	// it is marshalled to math.MaxUint64
-	transactionNumber uint64 // Gets marshalled to disk.
+	transactionNumber uint64 // marshalled to disk.
 
 	// payload contains the marshalled update, which may be spread over multiple
 	// pages if it is large. If spread over multiple pages, the full payload
 	// can be assembled by appending the separate payloads together. To keep the
 	// full size of the page at under pageSize bytes, the payload should not
 	// be greater than (pageSize - 64 bytes).
-	payload []byte // Gets marshalled to disk.
+	payload []byte // marshalled to disk.
 }
 
-// Marshal marshals a page.
-func (p page) Marshal() ([]byte, error) {
+func (p page) size() int { return pageMetaSize + len(p.payload) }
+
+func (p *page) writeToNoChecksum(w io.Writer) error {
```

Comment:
Nontrivial function needs a comment

Reply:
Also, I think it could be named better. When you write a checksum, your checksum covers every page, not just the page that you are writing. So `writeTo` as operating on a page just simply should never have a checksum. We've built it so that the checksum is the very first bytes of a marshalled transaction (iirc), which means that the transaction itself can compute the checksum and then the pages don't actually need the checksum at all.

Where it gets confusing is that we keep everything as compact and page-aligned as possible. What I'd really like to see (in a fully optimized version of the wal) is that the first page knows it's payload has to be shorter than the other pages, and then the rest of the pages have (almost) the full 4096 bytes.

Each page after the first page still needs 2 things:

1. A pointer to the next page (8 bytes)
2. One byte indicating that they aren't the first page, they are a secondary page

But, at least for a first pass this complicated the code a lot to have pages be different structures depending on their status as first or not-first (and, in the designs we had when we started, the last page was actually also special, so we would have needed 3 structs), so we kept it cleaner and figured we could optimize it further after initial implementation..

Reply:
The way I'm thinking we could do this that would be clean in code is that the `Transaction` type fully encompasses the first page, and then the `page` type encompasses all later pages. I don't think that's as much of a code transition as it would be to have all of a `Transaction`, a `firstPage`, and a `page`

Reply:
Probably best to punt this transition to the next PR